### PR TITLE
Tweak the run-e2e script to work on Windows

### PR DESCRIPTION
**Pull request by Tobbe**
# Why?

On Windows, running `./tasks/run-e2e` never gets past starting the RW dev server. 

image

That's as far as it gets. After that nothing more happens.

# How?

Runs the dev server async and instead waits for the :8910 and :8911 ports to answer on ping to know when to continue to the next step.

Additionally the script now kills the dev server after the user manually closes the Cypress window, so that the run-e2e script finishes. Without that final step the process wouldn't exit.

I'm creating this as a Draft, because I have no idea how this affects things on Linux. @thedavidprice could you please test this, see if it still works as expected on your machine?

BTW, is the run-e2e script supposed to, you know, actually run the e2e tests as well? Or just start Cypress and leave it up to the user to run (click) the tests they want to run? @dac09 I believe you wrote this script (at least translated it to nodejs). What's the expected behaviour (<- 'u' for you Danny 😉) here?

### CODE REVIEW

**Comment by thedavidprice** [comment]
> @thedavidprice could you please test this, see if it still works as expected on your machine?

^^ on my lists 👍

> BTW, is the run-e2e script supposed to, you know, actually run the e2e tests as well? 

^^ no, it builds the project and starts the dev server and Cypress. Starting the Cypress tests is a separate step. Effectively, this script handles everything in the GitHub Action workflow prior to the Cypres step (but for local testing). Does that make sense?

**Comment by thedavidprice** [comment]
Script ran successfully, launched Cypress, and I was able to complete the Cypress tests.

Script seemed almost 2x slower than normal. And Cypress also seemed much slower -- I encountered a few timeout errors. Could this be because your changes create more background processes?

**Comment by Tobbe** [comment]
> ^^ no, it builds the project and starts the dev server and Cypress. Starting the Cypress tests is a separate step. Effectively, this script handles everything in the GitHub Action workflow prior to the Cypres step (but for local testing). 

Right

> Does that make sense?

Not really 😆

I think, with a name like `run-e2e`, it should actually run the tests too. The GitHub action runs the script with the `--no-start` flag, so if we made starting the cypress tests part of that flag, no changes to the GH action would be required. WDYT?

> Script seemed almost 2x slower than normal. And Cypress also seemed much slower -- I encountered a few timeout errors. Could this be because your changes create more background processes?

`runDevServerInBackground` used to be `execa.sync`, but because of the way it was called, I don't actually think the next task waited for it to finish - effectively making it async anyway. Now that it waits for the dev server to start replying to requests before continuing the script is truly synchronous. If I'm right it previously started the dev server and the cypress tasks in parallel, causing a race condition where the tests could potentially start (and fail) before the dev server was fully up and running. But because of the human element of having to manually launch the tests the dev server often won the race, and no issue was detected. On a Windows machine, where starting the dev server is much slower, this race condition is more likely to be a problem.

All of that, however, does not explain why the actual tests would be slower. For that I have no good answer 🙁 I did just push a small tweak that possibly creates one less background process, but I'm not sure how much of a difference that makes.

**Comment by dac09** [comment]
> BTW, is the run-e2e script supposed to, you know, actually run the e2e tests as well? Or just start Cypress and leave it up to the user to run (click) the tests they want to run? @dac09 I believe you wrote this script (at least translated it to nodejs). What's the expected behaviour (<- 'u' for you Danny 😉) here?

No preference from me on automatically running the tests @Tobbe. I just maintained the behaviour from previous iterations of the script - if it aint broke dont fix it and all of that.

**Comment by thedavidprice** [comment]
> I think, with a name like run-e2e, it should actually run the tests too. The GitHub action runs the script with the --no-start flag, so if we made starting the cypress tests part of that flag, no changes to the GH action would be required. WDYT?

^^ tasks/ are mostly for local utilities/tools (in my interpretation), which may also be used for GitHub Actions. I use run-2e2 a _lot_ locally and would like to keep the behavior as is — there are two (soon to be 4) different Cypress tests, which I don't want to start automatically or always run all.

Understood about timing the processes, which might have been working on Mac but not Windows previously. As long as Peter and Danny end up with expected performance, disregard me. I (finally) have a new computer coming.

File: tasks/run-e2e [hunk 1]
```diff
@@ -6,6 +6,8 @@ const path = require('path')
 
 const execa = require('execa')
 const fs = require('fs-extra')
+const kill = require('tree-kill');
```
Comment (dac09) [inline tasks/run-e2e:9]: The name of this module makes me sad, especially on redwood 🙊

**Comment by dac09** [comment]
>Understood about timing the processes, which might have been working on Mac but not Windows previously. As long as Peter and Danny end up with expected performance, disregard me. I (finally) have a new computer coming.

Subtle ⭐. Sure @thedavidprice, I'll test it!

**Comment by Tobbe** [comment]
> ^^ tasks/ are mostly for local utilities/tools (in my interpretation), which may also be used for GitHub Actions. I use run-2e2 a _lot_ locally and would like to keep the behavior as is — there are two (soon to be 4) different Cypress tests, which I don't want to start automatically or always run all.

👍

**Comment by dac09** [comment]
Running perfectly fine for me locally. However 2 things to note:

1. I get these warnings from the api server (from `rw dev`):
```terminal
d to [process]. Use emitter.setMaxListeners() to increase limit
api | (node:28150) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGVTALRM listeners added to [process]. Use emitter.setMaxListeners() to increase limit
```

2. One I cancel the task I get an error
```terminal
^C/Users/dac09/Code/redwood/node_modules/execa/index.js:174
		err = new Error(`Command failed: ${joinedCmd}${output}`);
		      ^

Error: Command failed: yarn cypress open --env RW_PATH=/var/folders/6w/z1d6fjgd17g94qr6dcx31c5w0000gn/T/redwood-e2e/2021-04-13T16-17-01.455Z
    at makeError (/Users/dac09/Code/redwood/node_modules/execa/index.js:174:9)
    at Function.module.exports.sync (/Users/dac09/Code/redwood/node_modules/execa/index.js:338:15)
    at runCypress (/Users/dac09/Code/redwood/tasks/run-e2e:128:9)
    at /Users/dac09/Code/redwood/tasks/run-e2e:196:11
    at processTicksAndRejections (node:internal/process/task_queues:94:5) {
  code: null,
  stdout: null,
  stderr: null,
  failed: true,
  signal: 'SIGINT',
  cmd: 'yarn cypress open --env RW_PATH=/var/folders/6w/z1d6fjgd17g94qr6dcx31c5w0000gn/T/redwood-e2e/2021-04-13T16-17-01.455Z',
  timedOut: false
}
``` 
Like chill dood, I'm cancelling it intentionally.

**Comment by Tobbe** [comment]
> Once I cancel the task (ctrl+c) I get an error

What if you close (all) the cypress window(s)? Do you still get that error?

**Comment by dac09** [comment]
> > Once I cancel the task (ctrl+c) I get an error
> 
> What if you close (all) the cypress window(s)? Do you still get that error?

Nope, no error once I close the main cypress "test runner" window

**Comment by thedavidprice** [comment]
@Tobbe in light of #2228, what do you suggest we do with this one?

**Comment by Tobbe** [comment]
> @Tobbe in light of #2228, what do you suggest we do with this one?

I don't know if it's because of how things work on Windows, or if it's just because my machine is slow, but I need this for the script to work. And if some other Windows based developer comes along I think it's a greater chance they'll help with making the actual tests run if they can at least get this script running without too much trouble. Furthermore I think my addition of the wait-on package makes for a more robust solution for everyone.

With that said, With that said, I don't like if this makes it slower for you (or anyone else) or if it causes error messages in the console like @dac09 mentioned.

So, personally, I'd like us to benchmark this compared to what we have now, and if we find it's much slower, see what we can do about that. And also, see if we can make it nice regarding that error message. But both those things are things I'd need help with from someone on a Mac.

Finally, if you don't think it's worth the time and effort I won't be mad if we just close this. I still know where to find it if I ever feel like picking it up again.

**Comment by thedavidprice** [comment]
Super helpful @Tobbe And agreed. To be clear, I am _terrified_ by the fact our E2E doesn't run on Windows. I'm just not convinced we can get there with Cypress.

Putting this on my list to experiment with this week. Will try to benchmark and report back.

**Comment by thedavidprice** [comment]
Likely moving E2E to Playwright. Closing this out.

**Pull request closed without merging**